shell/platform/windows/game_pad_winuwp.cc

```diff
@@ -0,0 +1,228 @@
+  for (auto gamepad : gamepads) {
+    enumerated_game_pads_.push_back(gamepad);
+  }
+}
+
+const winrt::Windows::Gaming::Input::Gamepad* GamepadWinUWP::GetLastGamepad() {
+  winrt::Windows::Gaming::Input::Gamepad* gamepad = nullptr;
+
+  if (!enumerated_game_pads_.empty()) {
+    gamepad = &enumerated_game_pads_.back();
+  }
+
+  return gamepad;
+}
+
+bool isValid(double value) {
+  if (value > 0.1 || value < -0.1) {
+    return true;
+  }
+  return false;
+}
+
+void GamepadWinUWP::Process() {
+  {
+    const std::lock_guard<std::mutex> lock(gamepad_mutex_);
+
+    if (current_game_pad_ == nullptr) {
+      return;
+    }
+
+    last_reading_ = current_game_pad_->GetCurrentReading();
+  }
+
+  int exitComboPressed = 0;
+
+  namespace gi = winrt::Windows::Gaming::Input;
+
+  GamepadButtonPressedInternal(last_reading_.Buttons);
+
+  // work out which values have changed since the last reading.
+  if (last_reading_.LeftThumbstickX != left_stick_x_value_) {
+    left_stick_x_value_ = last_reading_.LeftThumbstickX;
+  }
+
+  if (last_reading_.LeftThumbstickY != left_stick_y_value_) {
+    left_stick_y_value_ = last_reading_.LeftThumbstickY;
+  }
+
+  if (isValid(left_stick_x_value_) || isValid(left_stick_y_value_)) {
+    RaiseLeftStickMoved(left_stick_x_value_, left_stick_y_value_);
+  }
+
+  if (last_reading_.RightThumbstickX != right_stick_x_value_) {
+    right_stick_x_value_ = last_reading_.RightThumbstickX;
+  }
+
+  if (last_reading_.RightThumbstickY != right_stick_y_value_) {
+    right_stick_y_value_ = last_reading_.RightThumbstickY;
+  }
+
+  if (isValid(right_stick_x_value_) || isValid(right_stick_y_value_)) {
+    RaiseRightStickMoved(right_stick_x_value_, right_stick_y_value_);
+  }
+
+  // Raise the requisit events based on what's changed since the last reading
+  // was taken.
+  if (last_reading_.LeftTrigger != 0 &&
+      last_reading_.LeftTrigger != left_trigger_value_) {
+    left_trigger_value_ = last_reading_.LeftTrigger;
+    RaiseLeftTriggerMoved(left_trigger_value_);
+  }
+
+  if (last_reading_.RightTrigger != 0 &&
+      last_reading_.RightTrigger != right_trigger_value_) {
+    right_trigger_value_ = last_reading_.RightTrigger;
+    RaiseRightTriggerMoved(right_trigger_value_);
+  }
+}
+
+void GamepadWinUWP::GamepadButtonPressedInternal(
+    winrt::Windows::Gaming::Input::GamepadButtons state) {
+  namespace wgi = winrt::Windows::Gaming::Input;
+
+  static const wgi::GamepadButtons AllButtons[] = {
+      wgi::GamepadButtons::A,         wgi::GamepadButtons::B,
+      wgi::GamepadButtons::DPadDown,  wgi::GamepadButtons::DPadLeft,
+      wgi::GamepadButtons::DPadRight, wgi::GamepadButtons::DPadUp};
+
+  for (const auto e : AllButtons) {
+    // if button is pressed
+    if ((e & state) == e) {
+      // if we have not sent a pressed already
+      if ((pressed_buttons_ & e) != e) {
+        // send pressed callback
+        if (button_pressed_callback_ != nullptr) {
+          button_pressed_callback_(e);
+        }
```

Comment:
Here and below: consider merging the outer two conditionals to reduce nesting.

Reply:
Fixed in next commit.